File: test/lib/vbrain/sources/text_test.rb
require "test_helper"
require "vbrain/sources"

class SourcesTextTest < Minitest::Test
  def test_detect_md_file
    with_tmpdir do |dir|
      f = File.join(dir, "foo.md")
      File.write(f, "# hi\n")
      assert VBrain::Sources::Text.detect?(f)
    end
  end

  def test_detect_txt_file
    with_tmpdir do |dir|
      f = File.join(dir, "foo.txt")
      File.write(f, "hello\n")
      assert VBrain::Sources::Text.detect?(f)
    end
  end

  def test_detect_extensionless_utf8_text
    with_tmpdir do |dir|
      f = File.join(dir, "notes")
      File.write(f, "some text\n")
      assert VBrain::Sources::Text.detect?(f)
    end
  end

  def test_detect_rejects_binary
    with_tmpdir do |dir|
      f = File.join(dir, "blob.bin")
      File.binwrite(f, "PK\x03\x04binarystuff\x00\xff\xfe")
      refute VBrain::Sources::Text.detect?(f)
    end
  end

  def test_detect_rejects_directory
    with_tmpdir do |dir|
      refute VBrain::Sources::Text.detect?(dir)
    end
  end

  def test_extract_writes_passthrough_content
    with_tmpdir do |dir|
      src = File.join(dir, "in.md")
      out = File.join(dir, "out.txt")
      File.write(src, "Olá Mundo\n")
      VBrain::Sources::Text.extract(src, out)
      assert_equal "Olá Mundo\n", File.read(out)
    end
  end

  def test_kind_key
    assert_equal "text", VBrain::Sources::Text.kind_key
  end
end

class SourcesDispatcherTest < Minitest::Test
  def test_detect_returns_text_for_md
    with_tmpdir do |dir|
      f = File.join(dir, "x.md")
      File.write(f, "# x")
      assert_equal VBrain::Sources::Text, VBrain::Sources.detect(f)
    end
  end

  def test_detect_returns_nil_for_binary
    with_tmpdir do |dir|
      f = File.join(dir, "b.bin")
      File.binwrite(f, "\x00\xff\x00\xff")
      assert_nil VBrain::Sources.detect(f)
    end
  end

  def test_for_lookup_by_kind
    assert_equal VBrain::Sources::Text, VBrain::Sources.for("text")
    assert_nil VBrain::Sources.for("nonexistent")
  end

  def test_kinds_includes_text
    assert_includes VBrain::Sources.kinds, "text"
  end
end

File: internal/sources/text_test.go
package sources_test

import (
	"os"
	"path/filepath"
	"testing"

	"github.com/virtual360-io/vbrain/internal/sources"
)

func writeTmp(t *testing.T, name string, data []byte) string {
	t.Helper()
	p := filepath.Join(t.TempDir(), name)
	if err := os.WriteFile(p, data, 0o644); err != nil {
		t.Fatal(err)
	}
	return p
}

func TestTextDetectMdFile(t *testing.T) {
	if !(sources.Text{}).Detect(writeTmp(t, "foo.md", []byte("# hi\n"))) {
		t.Fatal("should detect .md")
	}
}

func TestTextDetectTxtFile(t *testing.T) {
	if !(sources.Text{}).Detect(writeTmp(t, "foo.txt", []byte("hello\n"))) {
		t.Fatal("should detect .txt")
	}
}

func TestTextDetectExtensionlessUtf8(t *testing.T) {
	if !(sources.Text{}).Detect(writeTmp(t, "notes", []byte("some text\n"))) {
		t.Fatal("should detect extensionless utf8 text")
	}
}

func TestTextDetectRejectsBinary(t *testing.T) {
	if (sources.Text{}).Detect(writeTmp(t, "blob.bin", []byte("PK\x03\x04binarystuff\x00\xff\xfe"))) {
		t.Fatal("should not detect a binary")
	}
}

func TestTextDetectRejectsDirectory(t *testing.T) {
	if (sources.Text{}).Detect(t.TempDir()) {
		t.Fatal("should not detect a directory")
	}
}

func TestTextExtractWritesPassthrough(t *testing.T) {
	dir := t.TempDir()
	src := filepath.Join(dir, "in.md")
	out := filepath.Join(dir, "out.txt")
	if err := os.WriteFile(src, []byte("Olá Mundo\n"), 0o644); err != nil {
		t.Fatal(err)
	}
	if err := (sources.Text{}).Extract(src, out, sources.RawInfo{}); err != nil {
		t.Fatal(err)
	}
	if b, _ := os.ReadFile(out); string(b) != "Olá Mundo\n" {
		t.Fatalf("got %q", b)
	}
}

func TestTextKindKey(t *testing.T) {
	if (sources.Text{}).KindKey() != "text" {
		t.Fatal("kind_key should be text")
	}
}

func TestDispatcherDetectReturnsTextForMd(t *testing.T) {
	s := sources.Detect(writeTmp(t, "x.md", []byte("# x")))
	if s == nil || s.KindKey() != "text" {
		t.Fatalf("got %v", s)
	}
}

func TestDispatcherDetectReturnsNilForBinary(t *testing.T) {
	if s := sources.Detect(writeTmp(t, "b.bin", []byte("\x00\xff\x00\xff"))); s != nil {
		t.Fatalf("got %v, want nil", s)
	}
}

func TestDispatcherForLookupByKind(t *testing.T) {
	if s := sources.For("text"); s == nil || s.KindKey() != "text" {
		t.Fatalf("For(text) = %v", s)
	}
	if s := sources.For("nonexistent"); s != nil {
		t.Fatalf("For(nonexistent) = %v, want nil", s)
	}
}

func TestDispatcherKindsIncludesText(t *testing.T) {
	found := false
	for _, k := range sources.Kinds() {
		if k == "text" {
			found = true
		}
	}
	if !found {
		t.Fatal("Kinds() should include text")
	}
}
